enterprise/coderd/coderd.go

```diff
@@ -2,48 +2,218 @@ package coderd
+// New constructs an Enterprise coderd API instance.
+// This handler is designed to wrap the AGPL Coder code and
+// layer Enterprise functionality on top as much as possible.
+func New(ctx context.Context, options *Options) (*API, error) {
+	if options.EntitlementsUpdateInterval == 0 {
+		options.EntitlementsUpdateInterval = 10 * time.Minute
+	}
+	if options.Keys == nil {
+		options.Keys = Keys
+	}
+	ctx, cancelFunc := context.WithCancel(ctx)
+	api := &API{
+		AGPL:    coderd.New(options.Options),
+		Options: options,
+
+		auditLogs:              codersdk.EntitlementNotEntitled,
+		cancelEntitlementsLoop: cancelFunc,
+	}
+	oauthConfigs := &httpmw.OAuth2Configs{
+		Github: options.GithubOAuth2Config,
+		OIDC:   options.OIDCConfig,
+	}
+	apiKeyMiddleware := httpmw.ExtractAPIKey(options.Database, oauthConfigs, false)
+
+	api.AGPL.APIHandler.Group(func(r chi.Router) {
+		r.Get("/entitlements", api.entitlements)
+		r.Route("/licenses", func(r chi.Router) {
+			r.Use(apiKeyMiddleware)
+			r.Post("/", api.postLicense)
+			r.Get("/", api.licenses)
+			r.Delete("/{id}", api.deleteLicense)
+		})
+	})
+
+	err := api.updateEntitlements(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("update entitlements: %w", err)
+	}
+	api.closeLicenseSubscribe, err = api.Pubsub.Subscribe(pubSubEventLicenses, func(ctx context.Context, message []byte) {
+		_ = api.updateEntitlements(ctx)
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("subscribe to license updates: %w", err)
+	}
+	go api.runEntitlementsLoop(ctx)
+
+	return api, nil
+}
+
+type Options struct {
+	*coderd.Options
+
+	EntitlementsUpdateInterval time.Duration
+	Keys                       map[string]ed25519.PublicKey
+}
+
+type API struct {
+	AGPL *coderd.API
+	*Options
+
+	closeLicenseSubscribe  func()
+	cancelEntitlementsLoop func()
+	mutex                  sync.RWMutex
+	hasLicense             bool
+	activeUsers            codersdk.Feature
+	auditLogs              codersdk.Entitlement
+}
+
+func (api *API) Close() error {
+	api.closeLicenseSubscribe()
+	api.cancelEntitlementsLoop()
+	return api.AGPL.Close()
+}
+
+func (api *API) runEntitlementsLoop(ctx context.Context) {
+	ticker := time.NewTicker(api.EntitlementsUpdateInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+		err := api.updateEntitlements(ctx)
+		if err != nil {
+			api.Logger.Warn(ctx, "failed to get feature entitlements", slog.Error(err))
+			continue
+		}
+	}
+}
+
+func (api *API) updateEntitlements(ctx context.Context) error {
+	licenses, err := api.Database.GetUnexpiredLicenses(ctx)
+	if err != nil {
+		return err
+	}
+	api.mutex.Lock()
+	defer api.mutex.Unlock()
+	now := time.Now()
+	auditLogs := api.auditLogs
+	for _, l := range licenses {
+		claims, err := validateDBLicense(l, api.Keys)
+		if err != nil {
+			api.Logger.Debug(ctx, "skipping invalid license",
+				slog.F("id", l.ID), slog.Error(err))
+			continue
+		}
+		api.hasLicense = true
```

Comment:
this construction fails to disable `api.hasLicense`, `api.activeUsers.Entitlement` etc if a user deletes their license or it expires.  Once they are entitled, we will continue to say they are entitled forever.

Reply:
Good catch. Will fix.